File: 20/main.ts
function isValid(s: string): boolean {
  let stack: string[] = [];
  for (let char of s) {
    console.log(stack);
    console.log(char);
    if (char === "(" || char === "[" || char === "{") stack.push(char);
    else if (char === ")" && stack[stack.length - 1] !== "(") return false;
    else if (char === "]" && stack[stack.length - 1] !== "[") return false;
    else if (char === "}" && stack[stack.length - 1] !== "{") return false;
    else stack.pop();
  }
  return stack.length === 0;
}

console.log(isValid("()"));
console.log(isValid("()[]{}"));
console.log(isValid("(}"));
console.log(isValid("({)}"));

File: 20/main.py
class Solution:
    def isValid(self, s: str) -> bool:
        """
        approach
        use a stack, if open, push onto stack
        if close, if not matching on stack, return False
        else, pop and continue
        return True if ended
        """
        stack = []
        for ch in s:
            if ch in ("(", "{", "["):
                stack.append(ch)
            elif ch == ")":
                if len(stack) > 0 and stack[-1] == "(":
                    stack.pop()
                else:
                    return False
            elif ch == "}":
                if len(stack) > 0 and stack[-1] == "{":
                    stack.pop()
                else:
                    return False
            elif ch == "]":
                if len(stack) > 0 and stack[-1] == "[":
                    stack.pop()
                else:
                    return False
        return len(stack) == 0
